utils/logalloc.cc

```diff
@@ -817,16 +833,20 @@ class segment_store {
     }
     void free_segment(segment *seg) noexcept {
         seg->~segment();
-        ::free(seg);
+        _backend->free_segment_memory(seg);
     }
     size_t max_segments() const noexcept {
-        return (_layout.end - _segments_base) / segment::size;
+        return (_backend->memory_layout().end - _backend->segments_base()) / segment::size;
     }
     bool can_allocate_more_segments() const noexcept {
-        return memory::stats().free_memory() >= non_lsa_reserve + segment::size;
+        if (_backend->overlaps_with_non_lsa_memory()) {
+            return _backend->free_memory() >= non_lsa_reserve + segment::size;
+        } else {
+            return _backend->free_memory() >= segment::size;
+        }
```

Comment:
Should we push this logic to the backend instead of switching on some boolean?

I don't understand this completely.

It's better to expose a constant as a non-virtual function (as a micro optimization)